CSV Generator Tool

davidvossel opened the pull request:
The new `csv-generator` tool generates the KubeVirt Operator Lifecycle Marketplace CSV file from source and is integrated into our `make generate && make manifests` built flow. 

This tool is also included in the virt-operator container image to allow external tools, like the Hyper Converged Operator (HCO), to generate the KubeVirt CSV file from a build artifact associated with a specific KubeVirt release. The HCO will be using this functionality in order to make its own custom releases that are an aggregate of many components (kubevirt, cdi, network-addons, etc...). 

Example: Using csv-generator to generate a CSV file from a virt-operator container. The container in this example is just one I pushed from my dev environment.
```
export CSV_GENERATOR_IMAGE=docker.io/vossel/virt-operator:cnv-generate-test-v1
export CSV_OUTPUT_FILE=my-generated-csv.yaml

docker run -it --rm --entrypoint /usr/bin/csv-generator $CSV_GENERATOR_IMAGE --namespace=kubevirt --dockerPrefix=my-registry --operatorImageVersion=sha256@123 --pullPolicy=IfNotPresent --verbosity=2 --apiSha=sha256@123 --controllerSha=sha256@123 --handlerSha=sha256@123 --launcherSha=sha256@123 --csvVersion=0.0.1 --replacesCsvVersion=0.0.0 --kubeVirtVersion=0.0.1 > $CSV_OUTPUT_FILE
```

**Release note**:
```release-note
CSV generator tool (usr/bin/csv-generator) included in virt-operator container image. 
```

dankenigsberg commented on hack/generate.sh:
```diff
@@ -36,20 +37,42 @@ virtapi_version=$(getVersion ".VirtApiSha")
 ${KUBEVIRT_DIR}/tools/resource-generator/resource-generator --type=virt-handler --namespace={{.Namespace}} --repository={{.DockerPrefix}} --version="$virthandler_version" --pullPolicy={{.ImagePullPolicy}} --verbosity={{.Verbosity}} >${KUBEVIRT_DIR}/manifests/generated/virt-handler.yaml.in
 
+# The generation code for CSV requires a valid semver to be used.
+# But we're trying to generate a template for a CSV here from code
```
if so, maybe you should call this tool `csv-template-generator` ?

davidvossel replied:
well, not exactly.  This code comment is in relation only to how KubeVirt is utilizing the tool in our build flow. I'm referring to something here that's just specific to KubeVirt and not HCO's usage of csv-generator. 

KubeVirt is generating a template using the tool by replacing the arguments passed into the csv-generator with go template variables... So instead of passing in the actual operator's image sha, we pass in something like this `{.OperatorSha}`, which is later filled in using another build step. 

The HCO will actually be passing in real args to csv-generator though

dankenigsberg commented on hack/generate.sh:
```diff
@@ -36,20 +37,42 @@ virtapi_version=$(getVersion ".VirtApiSha")
+_fake_replaces_csv_version="1111.1111.1111"
+_fake_csv_version="2222.2222.2222"
+${KUBEVIRT_DIR}/tools/csv-generator/csv-generator --namespace={{.Namespace}} --dockerPrefix={{.DockerPrefix}} --operatorImageVersion="$virtoperator_version" --pullPolicy={{.ImagePullPolicy}} --verbosity={{.Verbosity}} --apiSha="$virtapi_rawsha" --controllerSha="$virtcontroller_rawsha" --handlerSha="$virthandler_rawsha" --launcherSha="$virtlauncher_rawsha" --kubevirtLogo={{.KubeVirtLogo}} --csvVer ...
+sed -i "s/$_fake_csv_version/{{.CsvVersion}}/g" ${KUBEVIRT_DIR}/manifests/generated/operator-csv.yaml.in
```
What is going to render the true CSV out of this template, filling the real value for `{{.CsvVersion}}`?
Can whatever that is doing it, simply pass `--csv-version` as an argument to `csv-generator` ?

davidvossel replied:
KubeVirt's `hack/build-manifests.sh` fills in the variables.

This isn't a model for how HCO needs to integrate with csv-generator necessarily though. Like my previous comment says, I envision the HCO using csv-generator with real arguments and not messing with all this templating logic like we're doing here. We'll see what makes sense for HCO though.

slintes commented on pkg/virt-operator/creation/csv/csv.go:
```diff
@@ -0,0 +1,350 @@
+# Virtualization extension for Kubernetes
+
+At its core, KubeVirt extends [Kubernetes](https://kubernetes.io) by adding
+additional virtualization resource types (especially the 'VirtualMachine' type) through
```
would be nice to keep the backticks here around VirtualMachine IMHO, because the description is rendered as markdown. Backticks can't be escaped, but you can do ....\` + "\`VirtualMachine\`" + \`....

davidvossel replied:
AH, yes. good idea. I'll give that a shot

slintes commented on pkg/virt-operator/creation/csv/csv.go:
```diff
@@ -0,0 +1,350 @@
+At its core, KubeVirt extends [Kubernetes](https://kubernetes.io) by adding
+additional virtualization resource types (especially the 'VirtualMachine' type) through
+[Kubernetes's Custom Resource Definitions API](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/).
+By using this mechanism, the Kubernetes API can be used to manage these 'VirtualMachine'
```
same here

davidvossel commented on hack/bazel-build.sh:
```diff
@@ -29,7 +29,7 @@ mkdir -p ${CMD_OUT_DIR}/virtctl
     --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 \
     --workspace_status_command=./hack/print-workspace-status.sh \
-    //cmd/...
+    //tools/csv-generator/... //cmd/...
```
It wasn't being built by bazel and therefore we couldn't put it into the container image.

slintes replied:
ah, fair point, sorry :)

davidvossel commented:
@slintes I think I addressed everything.

* restored "placeholder" namespace in CSV
* fixed backticks in CSV description
* filtered out creationTimestamp in deployments section of CSV

slintes commented:
/lgtm
/approve

davidvossel commented:
/test pull-kubevirt-e2e-k8s-1.11.0

slintes commented:
/lgtm
